### ISSUE TITLE
Using kubectl apply updates deployment but leaves pods behind

### ISSUE THREAD

**Issue by johnparn** [kubernetes/kubernetes#39798]
**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

I haven't found any duplicates just a similar issue: #27721. In that particular case the problem arose when using `kubectl rolling-update`. In my case I see the same problem when using `kubectl apply -f deplyoment.yaml`. It's not happening all the time, quite seldom actually.

---

**Kubernetes version** (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.6", GitCommit:"e569a27d02001e343cb68086bc06d47804f62af6", GitTreeState:"clean", BuildDate:"2016-11-12T05:22:15Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.7", GitCommit:"92b4f971662de9d8770f8dcd2ee01ec226a6f6c0", GitTreeState:"clean", BuildDate:"2016-12-10T04:43:42Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Google Container Engine**:

**What happened**:

I ran `kubectl apply -f deployment.yaml` to update. The deployment is updated to the new version however the pods are still sitting on the old version.

The deployment file is quite simple:

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: desktop-web
  namespace: staging
spec:
  replicas: 2
  template:
    metadata:
      labels:
        app: desktop-web
    spec:
      containers:
      - image: eu.gcr.io/websites-000000/desktop-web:0.0.202
        imagePullPolicy: Always
        name: desktop-web
        ports:
        - containerPort: 80
        readinessProbe:
          httpGet:
            path: /index.html
            port: 80
          initialDelaySeconds: 30
          timeoutSeconds: 1
```

View of deployments that is updated to version 202

![deployments](https://cloud.githubusercontent.com/assets/1776929/21888344/284a058a-d8c4-11e6-85b6-3e4dc0140656.png)

View of pods, still on 200 some hours after the deployment

![pods](https://cloud.githubusercontent.com/assets/1776929/21888343/2849e564-d8c4-11e6-86bf-41c99c1ac272.png)

The pod statuses

![pod-statuses](https://cloud.githubusercontent.com/assets/1776929/21888345/284a6cc8-d8c4-11e6-9a30-4d204c33efc4.png)

**What you expected to happen**:

The expected outcome of the command above is to update the deployment to the latest version specified in the file _deployment.yaml_.

**How to reproduce it** (as minimally and precisely as possible):

Unfortunately this happens quite seldom and I don't know what triggers this. We are using CI with automated deployment of successfully built and tested containers. However I have experienced this once or twice before even when doing manual deployments.

**Workaround**:

In this state, where deployment is version 202 and the pods are version 200, if I change the version in the _deployment.yaml_ back to the same version as the pods, ie version 200 ,and apply, the deployment is set to version 200, as expected. Nothing is expected to happen with the pods in this case. Bumping the version to 202 in deployment.yaml and applying again has no effect on the pods. The deployment version is reported as 202 but pods are still version 200. 

However by changing some other value like _initialDelaySeconds_ in the deployment file and applying, the pods are actually updated to 202.

**Comment by kargakis** [comment]
The deployment is not properly cleaned up (lacks revisionHistoryLimit so I guess you have all 200 replica sets?). Probably a dupe of https://github.com/kubernetes/kubernetes/issues/29735 but if you can provide any logs from the controller manager that would be nice.

**Comment by johnparn** [comment]
@kargakis That is right on spot! I had about 200 replicasets lying. Now I have applied the revisionHistoryLimit and I may perhaps destroyed any evidence on the way then. But which controller did you want the logs from (I'm on Google Container engine)?

For anyone else running into this problem, this is what the _deployment.yaml_ looks like with the new  _revisionHistoryLimit_ attribute:

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: desktop-web
  namespace: production
spec:
  replicas: 2
  revisionHistoryLimit: 10
  template:
    metadata:
      labels:
        app: desktop-web
    spec:
      containers:
      - image: eu.gcr.io/websites-******/desktop-web:0.0.219
        imagePullPolicy: Always
        name: desktop-web
        ports:
        - containerPort: 80
        # Check if site is responding, else restart
        readinessProbe:
          httpGet:
            path: /index.html
            port: 80
          initialDelaySeconds: 30
          timeoutSeconds: 1
```

**Comment by kargakis** [comment]
Not sure how to retrieve logs from GCE but looking at the logs from the controller manager would verify that this is likely a dupe of https://github.com/kubernetes/kubernetes/issues/29735. The cleanup of the old replica sets doesn't matter.

**Comment by kargakis** [comment]
Should be fixed in 1.7
